### DISCUSSION TITLE
What is the maximum amount of monitors you can have connected? Does it depend on their resolution?

### DISCUSSION THREAD

**Discussion by Evernow** [NVIDIA/open-gpu-kernel-modules#396] [Q&A] [answered]
Hello. 

What is exactly what determines the maximum amount of monitors you can have connected? Is it an overall bandwidth budget shared across the GPU's display output connectors, or does each connector get its own dedicated bandwidth?

For example, there is the DSC case: https://nvidia.custhelp.com/app/answers/detail/a_id/5338/~/using-high-resolution%2Frefresh-rate-displays-with-vesa-display-stream

> When a display is connected to the GPU and is set to DSC mode, the GPU may use two internal heads to drive the display when the pixel rate needed to drive the display mode exceeds the GPU’s single head limit. This may affect your display topology when using multiple monitors.  For example if two displays with support for DSC are connected to a single GeForce GPU, all 4 internal heads will be utilized and you will not be able to use a third monitor with the GPU at the same time.

This makes it seem like when using DSC the bandwidth of two connectors is used for one connector, is this assumption accurate? If DSC is not used can you connect a monitor to each of the GPU's display outputs and use up their respective max supported resolution / refresh rate?

**Comment by aritger** [comment, marked as answer]
Hi.  In general, a single NVIDIA GPU can drive four monitors: the GPU has four "heads" (the engine that generates raster timings to be sent to a monitor).  This of course depends on what physical connectors are on your graphics card, though you can use DisplayPort MultiStream to connect multiple DisplayPort monitors to one DP connector on the graphics card.

Each head has to fetch pixels from video memory; the memory bandwidth available to the heads is shared across all heads, but it is finite, and varies by the configuration.  If you're stepping through the code, the "IsModePossible" path attempts to validate if there is sufficient memory bandwidth to satisfy the GPU-wide requested configuration.

Between the head and the monitor is another hardware block on the GPU called the "Output Resource" (OR) that translates the timings produced by the head to the protocol of the particular monitor (TMDS, DP, etc).  Display Stream Compression (DSC) is primarily about compressing the data in the protocol, so between the OR and the monitor.

The section you quote about using two heads in some configurations is a little bit different: we'll use DSC in a fair number of cases to compress the data between the OR and the monitor, but in some of the most extreme cases a single head cannot fetch and produce pixels fast enough to feed the OR (IIRC, the limit is somewhere around 1.3G pixels/second, but my numbers may be off).  Those are the cases where we need to use two heads.

The quoted text is describing NVIDIA's Windows driver.  The Linux driver does not yet implement using 2 heads in this way; we're working on it, but I don't have a firm ETA.

I hope that helps.

**Reply by Evernow** [reply]
Thank you very much for this explanation! Greatly appreciated!

So from what I'm gathering, 1.3 gigapixels is 1,300,000,000. That is the limit of one header, and if it goes above that it'll use two headers (on Windows) and then in this situation you will only be able to plug in two more monitors assuming both of them are under the 1.3G pixels / second limit of their respective assigned headers. 

Is this gigapixel per second figure simply found by multiplying the resolution times the refresh rate? So for 3840x2160x144 it would be 1,194,393,600 pixels/second (or 1.3G pixels/second). Does it being 8-bit or 10-bit color matter here?

Thank you very much by the way, your explanation helped a lot to explain what is actually happening.

**Reply by aritger** [reply]
I'm not completely familiar with all the limitations, but I think resolution times refresh rate is a close enough approximation for purposes of this thread.  I'm also not positive the 1.3G pixels/second number is correct, especially across all GPU generations.

I don't believe 8-bit versus 10-bit impacts the calculation: both require fetching 4 bytes from video memory.  But, I haven't been too involved in this project, so I don't know the details of the limits as well as I'd like.

**Comment by cars11** [comment]
so does this mean the 5000 series cards with dp 2.1 will be able to handle 3-4 monitors at 4k240hz if they are using dp 2.1?

**Reply by Lablace** [reply]
I am also curious about whether 50-series can drive up to 4 4K 240Hz monitors without DSC. For AMD, even the highest variant Pro W7900 can only drive 4 4K 120Hz using DSC, which is frustrating.
